src/linsolve/lsmr.jl

```diff
@@ -0,0 +1,214 @@
+            ζbar = -sbar * ζbar
+
+            # Update h, h_hat, x.
+            hbar = add!!(hbar, h, 1, -θbar * ρ / (ρold * ρbarold))
+            h = add!!(h, v, 1, -θnew / ρ)
+            x = add!!(x, hbar, ζ / (ρ * ρbar), 1)
+
+            ##############################################################################
+            ##
+            ## Estimate of ||r||
+            ##
+            ##############################################################################
+
+            # Apply rotation Qhat_{k,2k+1}.
+            βacute = chat * βdd
+            βcheck = -shat * βdd
+
+            # Apply rotation Q_{k,k+1}.
+            βhat = c * βacute
+            βdd = -s * βacute
+
+            # Apply rotation Qtilde_{k-1}.
+            θtildeold = θtilde
+            ρtildeold = hypot(ρdold, θbar)
+            ctildeold = ρdold / ρtildeold
+            stildeold = θbar / ρtildeold
+            θtilde = stildeold * ρbar
+            ρdold = ctildeold * ρbar
+            βd = -stildeold * βd + ctildeold * βhat
+
+            τtildeold = (ζold - θtildeold * τtildeold) / ρtildeold
+            τd = (ζ - θtilde * τtildeold) / ρdold
+            d += abs2(βcheck)
+            normr = sqrt(d + abs2(βd - τd) + abs2(βdd))
+
+            # Estimate ||A||.
+            normA2 += abs2(β)
+            normA = sqrt(normA2)
+            normA2 += abs2(α)
+
+            # Estimate cond(A).
+            maxrbar = max(maxrbar, ρbarold)
+            if length(fact) > 1
+                minrbar = min(minrbar, ρbarold)
+            end
+            condA = max(maxrbar, ρtemp) / min(minrbar, ρtemp)
+
+            ##############################################################################
+            ##
+            ## Test for convergence
+            ##
+            ##############################################################################
+
+            # Compute norms for convergence testing.
+            normAr = abs(ζbar)
+            normx = norm(x)
+
+            # Now use these norms to estimate certain other quantities,
+            # some of which will be small near a solution.
+            test1 = normr / normb
+            test2 = normAr / (normA * normr)
+            test3 = inv(condA)
+
+            t1 = test1 / (one(Tr) + normA * normx / normb)
+            rtol = alg.btol + alg.atol * normA * normx / normb
+            # The following tests guard against extremely small values of
+            # atol, btol or ctol.  (The user may have set any or all of
+            # the parameters atol, btol, conlim  to 0.)
+            # The effect is equivalent to the normAl tests using
+            # atol = eps,  btol = eps,  conlim = 1/eps.
+
+            if alg.verbosity > 2
+                msg = "LSMR linsolve in iter $topit; step $(length(fact)-1): "
+                msg *= "normres = "
+                msg *= @sprintf("%.12e", normr)
+                @info msg
+            end
+
+            if 1 + test3 <= 1
```

Comment:
Can this be re-expressed as `test3 <= eps(one(test3))` ? This is somewhat strange code, because you could easily think that this just means `test3 <= 0`, which would be a different check.